File: Algorithm_Rust/examples/1161. 最大层内元素和.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function maxLevelSum(root: TreeNode | null): number {
  let stack: TreeNode[] = [];
  if (!root) {
    return 0;
  }
  let res = 0;
  let maxSum = 0;
  let count = 0;
  stack.push(root);
  while (stack.length) {
    let len = stack.length;
    count += 1;
    let sum = 0;
    while (len) {
      let node = stack.shift();
      node?.left && stack.push(node.left);
      node?.right && stack.push(node.right);
      sum += node!.val;
      len -= 1;
    }
    console.log(sum, count);
    if (sum > maxSum) {
      res = count;
    }
  }
  return res;
}

File: Algorithm_Rust/examples/1161. 最大层内元素和.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::collections::VecDeque;
use std::i32;
use std::rc::Rc;
impl Solution {
    pub fn max_level_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if root.is_none() {
            return 0;
        }
        let mut res = 0;
        let mut count = 0;
        let mut max_sum = i32::MAX;
        let mut stack: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        if let Some(node) = root.as_ref() {
            stack.push_back(node.clone());
        }
        while stack.len() > 0 {
            count += 1;
            let mut sum = 0;
            let mut len = stack.len();
            while len > 0 {
                let node = stack.pop_front().unwrap();
                sum += node.borrow().val;
                if let Some(left) = node.borrow().left.as_ref() {
                    stack.push_back(left.clone());
                }
                if let Some(right) = node.borrow().right.as_ref() {
                    stack.push_back(right.clone());
                }
                len -= 1;
            }
            println!("{} {}", sum, count);
            if sum > max_sum {
                max_sum = sum;
                res = count;
            }
        }
        res
    }
}
